router/internal/batch/batch.go

```diff
@@ -0,0 +1,185 @@
+package batch
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"github.com/cespare/xxhash/v2"
+	"github.com/go-chi/chi/v5"
+	"github.com/goccy/go-json"
+	"github.com/wundergraph/cosmo/router/pkg/config"
+	"github.com/wundergraph/cosmo/router/pkg/otel"
+	ctrace "github.com/wundergraph/cosmo/router/pkg/trace"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+	"io"
+	"net/http"
+	"strconv"
+)
+
+const defaultBufioReaderSize = 4096
+
+func getFirstNonWhitespaceChar(r io.Reader, readerSize int) (*byte, *bufio.Reader, error) {
+	// This uses the default buffer of 4 kb
+	bufReader := bufio.NewReaderSize(r, readerSize)
+
+	for {
+		peeked, err := bufReader.Peek(1)
+		if err != nil {
+			if err == io.EOF {
+				return nil, bufReader, nil
+			}
+			return nil, nil, err
+		}
+
+		if len(peeked) == 0 {
+			return nil, bufReader, nil
+		}
+		peekByte := peeked[0]
+		switch peekByte {
+		// we check the characters based on this RFC https://datatracker.ietf.org/doc/html/rfc8259
+		// and also the array decode function in goccy/go-json (which is the library we used to decode)
+		case ' ', '\n', '\t', '\r':
+			bufReader.ReadByte()
+			continue
+		default:
+			return &peekByte, bufReader, nil
+		}
+	}
+}
+
+type HandlerOpts struct {
+	MaxEntriesPerBatch  int
+	MaxRoutines         int
+	HandlerSent         http.Handler
+	Tracer              trace.Tracer
+	ClientHeader        config.ClientHeader
+	BaseOtelAttributes  []attribute.KeyValue
+	RouterConfigVersion string
+}
+
+func Handler(handlerOpts HandlerOpts) http.Handler {
+	fn := func(w http.ResponseWriter, r *http.Request) {
+		firstChar, bufReader, err := getFirstNonWhitespaceChar(r.Body, defaultBufioReaderSize)
+		if err != nil {
+			http.Error(w, "failed to read request", http.StatusBadRequest)
+			return
+		}
+
+		// When the first non whitespace character is not
+		// an array start assume it's a non batched request
+		if firstChar == nil || *firstChar != '[' {
+			// if firstChar is nil we have downstream handle it
+			// which is the current behaviour
+			r.Body = io.NopCloser(bufReader)
+			handlerOpts.HandlerSent.ServeHTTP(w, r)
+			return
+		}
+
+		bodyBytes, err := io.ReadAll(bufReader)
+		if err != nil {
+			http.Error(w, "failed to read request body", http.StatusBadRequest)
+			return
+		}
+
+		var batchOperations []json.RawMessage
+		if err = json.Unmarshal(bodyBytes, &batchOperations); err != nil {
+			// If there is an error, it's likely a malformed json array
+			// as the start character is "["
+			http.Error(w, "unexpected request body", http.StatusBadRequest)
+			return
+		}
+
+		batchOperationsLength := len(batchOperations)
+
+		addTracing(r, bodyBytes, handlerOpts.ClientHeader, batchOperationsLength, handlerOpts.BaseOtelAttributes, handlerOpts.RouterConfigVersion)
+
+		// When a max batch limit has been specified
+		if batchOperationsLength > handlerOpts.MaxEntriesPerBatch {
+			http.Error(w, "unable to process request", http.StatusBadRequest)
```

Comment:
How does other implementations behave? Do they fail or do they also allow partial responses?

This error should be tracked via tracing.

Reply:
Will need to check this.

Reply:
Does not handle partials, just errors out. I made it similar.

Reply:
Alright